### PR TITLE
BCI-2450: Health check for MultiNode to ensure that node is not in syncing state

### PR DESCRIPTION
**Pull request by dhaidashenko**
This came up as a follow-up of testing on canary, by QA teams.

If an RPC node is newly setup, and still syncing the blockchain, it may not be ready for taking client requests. Currently, our MultiNode client marks this node as OutOfSync, and logs error statements in our logs, which is confusing for some NOPS.

Changes:
* Added separate node state `syncing` to differentiate it from outOfSyncStatte.
* Fixed reconnection bug for nodes with invalidChainID. We close ws connection, when transitioning node to this state, but do not reconnect.

### CODE REVIEW

File: common/client/node_fsm.go [hunk 1]
```diff
@@ -231,6 +240,21 @@ func (n *node[CHAIN_ID, HEAD, RPC]) transitionToUnreachable(fn func()) {
...
+	case nodeStateInvalidChainID:
+		n.declareInvalidChainID()
+	case nodeStateUnreachable:
+		n.declareUnreachable()
+	case nodeStateSyncing:
+		n.declareSyncing()
+	case nodeStateAlive:
+		n.declareAlive()
+	default:
+		panic(fmt.Sprintf("%#v state declaration is not impletemented", state))
```
Comment (samsondav) [inline common/client/node_fsm.go:254]:
```suggestion
		panic(fmt.Sprintf("%#v state declaration is not implemented", state))
```

File: common/client/node.go [hunk 2]
```diff
@@ -224,53 +224,81 @@ func (n *node[CHAIN_ID, HEAD, RPC]) start(startCtx context.Context) {
...
+	n.declareState(state)
 }
 
-// verify checks that all connections to eth nodes match the given chain ID
+// verifyChainID checks that connection to the node matches the given chain ID
 // Not thread-safe
-// Pure verify: does not mutate node "state" field.
-func (n *node[CHAIN_ID, HEAD, RPC]) verify(callerCtx context.Context) (err error) {
+// Pure verifyChainID: does not mutate node "state" field.
+func (n *node[CHAIN_ID, HEAD, RPC]) verifyChainID(callerCtx context.Context, lggr logger.Logger) nodeState {
```
Comment (samsondav) [inline common/client/node.go:234]: Why does this return a state now?

File: common/client/node_lifecycle.go [hunk 3]
```diff
@@ -414,6 +403,14 @@ func (n *node[CHAIN_ID, HEAD, RPC]) invalidChainIDLoop() {
...
 
 	lggr := logger.Named(n.lfcLog, "InvalidChainID")
+
+	// Need to redial since invalid chain ID nodes are automatically disconnected
+	state := n.createVerifiedConn(n.nodeCtx, lggr)
+	if state != nodeStateInvalidChainID {
+		n.declareState(state)
+		return
+	}
+
```
Comment (samsondav) [inline common/client/node_lifecycle.go:413]: Why was this added? We never had a bug with it before.

File: common/client/node_lifecycle_test.go [hunk 4]
```diff
@@ -471,21 +474,24 @@ func TestUnit_NodeLifecycle_outOfSyncLoop(t *testing.T) {
 			return node.State() == nodeStateUnreachable
 		})
 	})
-	t.Run("if fail to get chainID, transitions to invalidChainID", func(t *testing.T) {
+	t.Run("if fail to get chainID, transitions to unreachable", func(t *testing.T) {
```
Comment (samsondav) [inline common/client/node_lifecycle_test.go:477]: Why this change in logic?

**Review by samsondav — CHANGES_REQUESTED** [review]

File: common/client/node_lifecycle_test.go (hunk 4 above)
Comment (dhaidashenko) [reply common/client/node_lifecycle_test.go:477]: This way it's possible to unify reconnection behaviour and reuse it in multiple state transitions and makes it easier to add additional verification steps. 

Before the change, error handling on chainID verification error for `outOfSync` loop looked like this
```
if err := n.verify(n.nodeCtx); err != nil {
		lggr.Errorw(fmt.Sprintf("Failed to verify out-of-sync RPC node: %v", err), "err", err)
		n.declareInvalidChainID()
		return
}
```

It was not consistent with the rest of the loops and produced state transition that seems redundant `outOfSync` -> `invalidChainID` -> `unreachable` instead of `outOfSync` -> `unreachable`

File: common/client/node_lifecycle.go [hunk 5]
```diff
@@ -423,18 +420,78 @@ func (n *node[CHAIN_ID, HEAD, RPC]) invalidChainIDLoop() {
...
+				continue
+			}
+
+			lggr.Infow(fmt.Sprintf("Successfully verified RPC node. Node was syncing for %s", time.Since(syncingAt)), "nodeState", n.State())
 			n.declareAlive()
 			return
+		case <-time.After(zombieNodeCheckInterval(n.noNewHeadsThreshold)):
+			if n.nLiveNodes != nil {
+				if l, _, _ := n.nLiveNodes(); l < 1 {
+					lggr.Critical("RPC endpoint is still syncing, but there are no other available nodes. This RPC node will be forcibly moved back into the live pool in a degraded state")
```
Comment (prashantkumar1982) [inline common/client/node_lifecycle.go:489]: Is this helpful?
Do we know if a syncing node can operate even in a degraded state?
For example, if it is not caught upto latest heads, then it can't serve heads, or check for receipts, or verify transactions sent to it.
Isn't it better to just let it sync first?

File: common/client/node.go [hunk 6]
```diff
@@ -224,53 +224,81 @@ func (n *node[CHAIN_ID, HEAD, RPC]) start(startCtx context.Context) {
...
+
+// verifyConn - verifies that current connection is valid: chainID matches, and it's not syncing.
+// Returns desired state if one of the verifications fails. Otherwise, returns nodeStateAlive.
+func (n *node[CHAIN_ID, HEAD, RPC]) verifyConn(ctx context.Context, lggr logger.Logger) nodeState {
+	state := n.verifyChainID(ctx, lggr)
+	if state != nodeStateAlive {
+		return state
+	}
+
+	isSyncing, err := n.rpc.IsSyncing(ctx)
```
Comment (prashantkumar1982) [inline common/client/node.go:290]: Can we put an additional config setting here?
NodeIsSyncingEnabled = false[default]

This will allow us to skip IsSyncing() calls for chains/clients that don't reliably implement this functionality, or have bugs around this behavior.

**Review by prashantkumar1982 — COMMENTED** [review]
Looks pretty good. 
Just a couple comments.

File: docs/CHANGELOG.md [hunk 7]
```diff
@@ -25,6 +26,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   echo "Baz=Val" >> median.env  
   CL_MEDIAN_ENV="median.env"
   ```
+>>>>>>> develop
```
[the remark this answers is not in the document]
Comment (dhaidashenko) [reply docs/CHANGELOG.md]: yep, thanks

File: common/client/node_lifecycle_test.go [hunk 8]
```diff
@@ -837,6 +836,26 @@ func TestUnit_NodeLifecycle_unreachableLoop(t *testing.T) {
 
 		setupRPCForAliveLoop(t, rpc)
 
+		node.declareUnreachable()
+		tests.AssertEventually(t, func() bool {
+			return node.State() == nodeStateAlive
+		})
+	})
+	t.Run("on successful verification without isSyncing becomes alive", func(t *testing.T) {
```
Comment (prashantkumar1982) [inline common/client/node_lifecycle_test.go:844]: nit: This same description is repeated multiple times for different test cases

File: common/client/node_lifecycle.go [hunk 9]
```diff
@@ -483,14 +484,6 @@ func (n *node[CHAIN_ID, HEAD, RPC]) syncingLoop() {
 			lggr.Infow(fmt.Sprintf("Successfully verified RPC node. Node was syncing for %s", time.Since(syncingAt)), "nodeState", n.State())
 			n.declareAlive()
 			return
-		case <-time.After(zombieNodeCheckInterval(n.noNewHeadsThreshold)):
```
Comment (prashantkumar1982) [inline common/client/node_lifecycle.go]: What was the reason for removing this?
Are we now not forcing to use a syncing node if that's the only node?

File: common/client/node_lifecycle_test.go (hunk 8 above)
Comment (dhaidashenko) [reply common/client/node_lifecycle_test.go:844]: Yes, but each of them is located in different testing functions and test similar transition: from unreachable to alive, from invalidChainID to alive, etc.

File: common/client/node_lifecycle.go (hunk 9 above)
Comment (dhaidashenko) [reply common/client/node_lifecycle.go]: Following [your suggestion](https://github.com/smartcontractkit/chainlink/pull/11765#discussion_r1460542437), I've run some tests, and it turns out that RPC is not able to properly serve requests while it's syncing and marking such node as Alive leads to a lot of error noise.

File: common/client/node.go (hunk 2 above)
Comment (samsondav) [reply common/client/node.go:235]: yeah this makes sense

File: common/client/node_lifecycle.go (hunk 3 above)
Comment (samsondav) [reply common/client/node_lifecycle.go:413]: ah interesting. Good catch!

File: common/client/node_fsm_test.go [hunk 10]
```diff
@@ -106,3 +121,9 @@ func testTransition(t *testing.T, rpc *mockNodeClient[types.ID, Head], transitio
 
 	}
 }
+
+func TestNodeState_String(t *testing.T) {
+	for _, ns := range allNodeStates {
+		assert.NotContains(t, ns.String(), strconv.FormatInt(int64(ns), 10), "Expected node state to have readable name")
```
[the remark this answers is not in the document]
Comment (dhaidashenko) [reply common/client/node_fsm_test.go]: Added description to the test

**Review by samsondav — DISMISSED** [review]

**Review by prashantkumar1982 — APPROVED** [review]
